pep-0612.rst

```diff
@@ -0,0 +1,258 @@
+   def foo(x: typing.List[TParams]) -> None: ... # Rejected
+   def foo(x: typing.Callable[[int, str], TParams]) -> None: ... # Rejected
+
+Semantics
+^^^^^^^^^
+
+The inference rules for the return type of a function invocation whose signature contains a ``ParameterSpecification`` variable are analogous to those around evaluating ones with ``TypeVar``\ s. 
+
+.. code-block::
+
+   def foo(x: typing.Callable[TParams, int]) -> typing.Callable[TParams, str]: ...
+   def bar(a: str, b: bool) -> int: ...
+   f = foo(bar) # f should be inferred to have the same signature as bar, but returning str
+   f("A", True) # Accepted
+   f(a = "A", b = True) # Accepted
+   f("A", "A") # Rejected
+
+Just as with traditional ``TypeVars``\ , a user may include the same ``ParameterSpecification`` multiple times in the arguments of the same function, to indicate a dependency between multiple arguments.  In these cases a type checker may choose to solve to a common behavioral supertype (i.e. a set of parameters for which all of the valid calls are valid in both of the subtypes), but is not obligated to do so.
+
+.. code-block::
+
+   def foo(
+       x: typing.Callable[TParams, int], y: typing.Callable[TParams, int]
+   ) -> typing.Callable[TParams, bool]: ...
+
+   def x_int_y_str(x: int, y: str) -> int: ...
+   def y_int_x_str(y: int, x: str) -> int: ...
+   foo(x_int_y_str, x_int_y_str) # Must return (x: int, y: str) -> int
+   foo(x_int_y_str, y_int_x_str) # Could return (__a: int, __b: str) -> int 
+                                 # This works because both callables have types that are behavioral
+                                 # subtypes of Callable[[int, str], int]
+
+   def keyword_only_x(*, x: int) -> int: ...
+   def keyword_only_y(*, y: int) -> int: ...
+   foo(keyword_only_x, keyword_only_y) # Must be rejected
+
+Use in ``Generic`` Classes
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Just as with normal ``TypeVar``\ s, ``ParameterSpecification``\ s can be used to make generic classes as well as generic functions. These ought to be able to be mixed with normal ``TypeVar``\ s. This should also be made to work with protocols in the same manner.
+
+The components of a ``ParameterSpecification``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+A ``ParameterSpecification`` captures both positional and keyword accessible parameters, but there unfortunately is no object in the runtime that captures both of these together. Instead, we are forced to separate them into ``*args`` and ``**kwargs``\ , respectively. This means we need to be able to split apart a single ``ParameterSpecification`` into these two components, and then bring them back together into a call.  To do this, we introduce ``TParams.args`` to represent the tuple of positional arguments in a given call and ``TParams.kwargs`` to represent the corresponding ``Mapping`` of keywords to values. These operators can only be used together, as the annotated types for ``*args`` and ``**kwargs`` .
```

Comment:
What is the motivation to require `TParams.args` and `TParams.kwargs` to be used together.

Why is the example `def baz(*args: TParams.args) -> int` rejected?